zebra-state/src/service.rs

```diff
@@ -29,14 +34,103 @@ struct StateService {
     /// Holds data relating to finalized chain state.
     sled: FinalizedState,
     /// Holds data relating to non-finalized chain state.
-    _mem: NonFinalizedState,
+    mem: NonFinalizedState,
+    /// Blocks awaiting their parent blocks for contextual verification.
+    contextual_queue: QueuedBlocks,
+}
+
+#[derive(Debug, Error)]
+#[error("block is not contextually valid")]
+struct CommitError(#[from] ValidateContextError);
+
+#[derive(displaydoc::Display, Debug, Error)]
+enum ValidateContextError {
+    /// block.height is lower than the current finalized height
+    OrphanedBlock,
 }
 
 impl StateService {
+    const REORG_LIMIT: usize = 100;
+
     pub fn new(config: Config, network: Network) -> Self {
         let sled = FinalizedState::new(&config, network);
-        let _mem = NonFinalizedState::default();
-        Self { sled, _mem }
+        let mem = NonFinalizedState::default();
+        let contextual_queue = QueuedBlocks::default();
+
+        Self {
+            sled,
+            mem,
+            contextual_queue,
+        }
+    }
+
+    fn queue(&mut self, new: QueuedBlock) {
+        let parent_hash = new.block.header.previous_block_hash;
+
+        self.contextual_queue.queue(new);
+
+        if !self.contains(&parent_hash) {
+            return;
+        }
+
+        self.process_queued(parent_hash);
+
+        while self.mem.best_chain_len() > Self::REORG_LIMIT {
```

Comment:
Using MAX_BLOCK_REORG_HEIGHT fixes an off-by-one error in this check.

The coinbase maturity rule is 100 blocks, so the maximum reorganisation is 99 blocks.

```suggestion
        while self.mem.best_chain_len() > MAX_BLOCK_REORG_HEIGHT {
```